GVFS/GVFS.Common/Cleanup/GitCleanupQueue.cs

```diff
@@ -97,40 +53,30 @@ private void RunQueue()
         {
             while (true)
             {
-                lock (this.threadLock)
-                {
-                    if (this.queue.Count == 0
-                        || !this.EnlistmentRootReady())
-                    {
-                        this.thread = null;
-                        return;
-                    }
-                }
-
-                lock (this.currentStepLock)
+                try
                 {
-                    this.queue.TryDequeue(out this.currentStep);
+                    this.queue.TryTake(out this.currentStep, Timeout.Infinite, this.cancellationToken.Token);
                 }
-
-                if (this.stopping || this.currentStep == null)
+                catch (OperationCanceledException)
                 {
-                    this.thread = null;
+                    // Only gets called when stop is requested
                     return;
                 }
 
-                try
-                {
-                    this.currentStep.Execute();
-                }
-                catch (Exception e)
+                if (this.EnlistmentRootReady())

```

Comment:
`if (!this.Stopping && this.EnlistmentRootReady())` to avoid a race where we `Stop()` our previous step, but also get a new step from the collection and start the new step.

Reply:
Since in stop we call:
this.cancellationToken.Cancel(); <- prevents new things to be taken from the queue
this.currentStep?.Stop(); <- if we have something currently we stop
I think we're good thread-wise there.

Looking at GitCleanUpStep I looks like me might have some race conditions there (unrelated).  I'll take a dive into that and update.